source/init/callback.h

```diff
@@ -0,0 +1,88 @@
+#pragma once
+
+#include "common/callback/callback.h"
+#include "common/common/logger.h"
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Init {
+
+/**
+ * A Caller notifies a Receiver when something has initialized. This works at two levels: first,
+ * each initialization target holds a Caller to notify the init manager's Receiver when the target
+ * has initialized. And second, the manager holds a Caller to notify its client's Receiver when all
+ * of its targets have initialized.
+ *
+ * Caller and Receiver are simple wrappers for their counterparts in Common::Callback with logging
+ * added to help debug initialization and destruction ordering issues that occasionally arise.
+ */
+class Caller : Logger::Loggable<Logger::Id::init> {
```

Comment:
Wait, why is there a `Caller`/`Receiver` defined here as well as earlier on? :) Can't we just have one abstract version of this?

Reply:
The `Caller`/`Receiver` in common/callback are meant to be as bare-bones as possible, with no logging included. Going on the assumption that we want safer callbacks everywhere, not just in initialization, I didn't want to bake logging as a requirement.

Reply:
We definitely want to converge these implementations IMHO; even if we have to use macro magic or somethign to do it. They are pretty complicated to reason about, having only a single one is the way to go.